cci-stats/pkg/service/service.go

```diff
@@ -0,0 +1,281 @@
+		}
+
+		jobs, err := fetchJobs(ctx, client, wf.ID)
+		if err != nil {
+			return fmt.Errorf("failed to fetch jobs: %w", err)
+		}
+		workflowsJobs[wf.ID] = jobs
+		workflowsByID[wf.ID] = wf
+	}
+
+	type testMetadata struct {
+		JobID    string
+		WFID     string
+		Metadata []*circleci.TestMetadata
+	}
+
+	slog.Debug("fetching test metadata", "pipeline", pline.ID, "number", pline.Number)
+	testPool := pool.NewWithResults[testMetadata]().
+		WithErrors().
+		WithFirstError().
+		WithMaxGoroutines(config.MaxConcurrentFetchJobs).
+		WithContext(ctx).
+		WithCancelOnError()
+	for wfID, jobs := range workflowsJobs {
+		for _, job := range jobs {
+			testPool.Go(func(ctx context.Context) (testMetadata, error) {
+				tm, err := fetchTestMetadata(ctx, config, client, job)
+				if err != nil {
+					return testMetadata{}, fmt.Errorf("failed to fetch test metadata: %w", err)
+				}
+				return testMetadata{
+					JobID:    job.ID,
+					WFID:     wfID,
+					Metadata: tm,
+				}, nil
+			})
+		}
+	}
+	metadata, err := testPool.Wait()
+	if err != nil {
+		return fmt.Errorf("failed to fetch test metadata: %w", err)
+	}
+
+	tx, err := dbConn.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to start transaction: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			tx.Rollback(ctx)
+		}
+	}()
+
+	slog.Info("inserting into DB", "pipeline", pline.ID, "number", pline.Number)
+	if err := tx.InsertPipeline(ctx, db.Pipeline{
+		ID:        pline.ID,
+		Number:    pline.Number,
+		Commit:    pline.Vcs.Revision,
+		Branch:    pline.Vcs.Branch,
+		CreatedAt: pline.CreatedAt,
+	}); err != nil {
+		return fmt.Errorf("failed to insert pipeline: %w", err)
+	}
+
+	slog.Info("inserting jobs", "pipeline", pline.ID)
+	for wfID, jobs := range workflowsJobs {
+		wf := workflowsByID[wfID]
+		for _, job := range jobs {
+			if err := tx.InsertJob(ctx, db.Job{
+				ID:           fmt.Sprintf("%s/%s", wf.ID, job.ID),
+				PipelineID:   pline.ID,
+				WorkflowID:   wfID,
+				WorkflowName: wf.Name,
+				Number:       job.JobNumber,
+				Name:         job.Name,
+				Status:       job.Status,
+				StartedAt:    job.StartedAt,
+				StoppedAt:    job.StoppedAt,
+			}); err != nil {
+				return fmt.Errorf("failed to insert workflow: %w", err)
+			}
+		}
+	}
+
+	slog.Info("inserting test metadata", "pipeline", pline.ID, "number", pline.Number)
+	for _, tm := range metadata {
+		for _, m := range tm.Metadata {
+			if _, err := tx.InsertTestResult(ctx, db.TestResult{
+				JobID:   fmt.Sprintf("%s/%s", tm.WFID, tm.JobID),
+				Name:    m.Name,
+				Runtime: m.RunTime,
+				Status:  m.Result,
+				Message: m.Message,
+			}); err != nil {
+				return fmt.Errorf("failed to insert test result: %w", err)
+			}
+		}
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+func fetchPipelines(ctx context.Context, config config.Config, cutoff time.Time, client *circleci.Client) ([]*circleci.Pipeline, error) {
```

Comment:
nit/idea: this could potentially be used to fetch the `LastPipeline`, thus simplifying reducing the db interface

Reply:
Also, another idea, this `cutoff` feels like it could perhaps be superseded by a pagination token.

Reply:
`LastPipeline` is used to determine what the latest pipeline in the database is. The `createdAt` value from that pipeline is then used as the `cutoff` value. When the build with the `createdAt` value matching the `cutoff` value is found, the application stops collecting pipelines and starts collecting other data (e.g. tests, workflows, etc.). Since the local value matters here, we'll still need the DB interface to expose a getter for its latest pipeline.

How would I use the pagination token here? There's no way to know how many pages we need to read in advance, and the token itself is returned by the CCI API when you make a request for a pipeline.